src/container/srv_target.c

```diff
@@ -1142,146 +1160,278 @@ ds_cont_tgt_epoch_discard_aggregator(crt_rpc_t *source, crt_rpc_t *result,
-	uint64_t				 i;
-	int					 rc;
+	struct cont_snap_args	*args = vin;
+	struct ds_cont_child	*cont;
+	int			 rc;
 
-	pool_child = ds_pool_child_lookup(in->tai_pool_uuid);
-	if (pool_child == NULL) {
-		D_ERROR(DF_CONT": Lookup pool child failed\n",
-			DP_CONT(in->tai_pool_uuid, in->tai_cont_uuid));
-		return -DER_NO_HDL;
+	rc = ds_cont_child_lookup(args->pool_uuid, args->cont_uuid, &cont);
+	if (rc != 0)
+		return rc;
+	if (args->snap_count == 0) {
+		if (cont->sc_snapshots != NULL) {
+			D_ASSERT(cont->sc_snapshots_nr > 0);
+			D_FREE(cont->sc_snapshots);
+			cont->sc_snapshots = NULL;
+		}
+	} else {
+		void	*buf;
+		size_t	 bufsize;
+
+		bufsize = args->snap_count * sizeof(*args->snapshots);
+		D_REALLOC(buf, cont->sc_snapshots, bufsize);
+		if (buf == NULL) {
+			rc = -DER_NOMEM;
+			goto out_cont;
+		}
+		memcpy(buf, args->snapshots, bufsize);
+		cont->sc_snapshots = buf;
 	}
+	cont->sc_snapshots_nr = args->snap_count;
+	cont->sc_aggregation_max = DAOS_EPOCH_MAX;
+	cont->sc_snapshots_ts = crt_hlc_get();
+out_cont:
+	ds_cont_child_put(cont);
+	return rc;
+}
 
-	rc = vos_cont_open(pool_child->spc_hdl, in->tai_cont_uuid,
-			   &cont_hdl);
-	if (rc != 0) {
-		D_ERROR(DF_CONT": Open container failed: %d\n",
-			DP_CONT(in->tai_pool_uuid, in->tai_cont_uuid), rc);
-		goto pool_child;
+int
+ds_cont_tgt_snapshots_update(uuid_t pool_uuid, uuid_t cont_uuid,
+			     uint64_t *snapshots, int snap_count)
+{
+	struct cont_snap_args	 args;
+
+	uuid_copy(args.pool_uuid, pool_uuid);
+	uuid_copy(args.cont_uuid, cont_uuid);
+	args.snap_count = snap_count;
+	args.snapshots = snapshots;
+	D_DEBUG(DB_TRACE, DF_UUID": refreshing snapshots %d\n",
+		DP_UUID(cont_uuid), snap_count);
+	return dss_thread_collective(cont_snap_update_one, &args, 0);
+}
+
+void
+cont_snapshots_refresh_ult(void *data)
+{
+	struct cont_snap_args	*args = data;
+	struct ds_pool		*pool;
+	int			 rc;
+
+	pool = ds_pool_lookup(args->pool_uuid);
+	if (pool == NULL) {
+		rc = -DER_NO_HDL;
+		goto out;
 	}
+	rc = cont_iv_snapshots_refresh(pool->sp_iv_ns, args->cont_uuid);
+	ds_pool_put(pool);
+out:
+	if (rc != 0)
+		D_WARN(DF_UUID": failed to refresh snapshots IV: rc %d;"
+			" Aggregation may not work correctly\n",
+			DP_UUID(args->cont_uuid), rc);
+	D_FREE(args);
+}
 
-	rc = vos_cont_query(cont_hdl, &cont_info);
-	if (rc != 0) {
-		D_ERROR(DF_CONT": Query container failed: %d\n",
-			DP_CONT(in->tai_pool_uuid, in->tai_cont_uuid), rc);
-		goto cont_close;
+int
+ds_cont_tgt_snapshots_refresh(uuid_t pool_uuid, uuid_t cont_uuid)
+{
+	struct cont_snap_args	*args;
+	int			 rc;
+
+	D_ALLOC(args, sizeof(*args));
+	if (args == NULL)
+		return -DER_NOMEM;
+	uuid_copy(args->pool_uuid, pool_uuid);
+	uuid_copy(args->cont_uuid, cont_uuid);
+	rc = dss_ult_create(cont_snapshots_refresh_ult, args,
+			    DSS_ULT_POOL_SRV, 0, 0, NULL);
+	if (rc != 0)
+		D_FREE(args);
+	return rc;
+}
+
+/** Currently set to DTX_AGG_THRESHOLD_AGE_LOWER **/
+#define DAOS_AGG_THRESHOLD 3600 /** seconds **/
```

Comment:
1 hour? Then I suppose there won't be aggregation triggered in testing, should we shrink it to 10 seconds or so?